src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -972,6 +974,28 @@ EnsureReplicationSettings(Oid relationId, char replicationModel)
 }
 
 
+/*
+ * EnsureRelationHasNoTriggers errors out if the given table has triggers on
+ * it. See also GetExplicitTriggerOids function's comment for the triggers this
+ * function errors out.
+ */
+static void
+EnsureRelationHasNoTriggers(Oid relationId)
+{
+	List *explicitTriggerIds = GetExplicitTriggerOids(relationId);
+
+	if (list_length(explicitTriggerIds) > 0)
```

Comment:
when I add ` && !IsReferenceTable(relationId)` here, and do the following:

```sql
CREATE TABLE ref_table(value int);

CREATE FUNCTION update_value() RETURNS trigger AS $update_value$
BEGIN
    NEW.value := value+1 ;
    RETURN NEW;
END;
$update_value$ LANGUAGE plpgsql;



CREATE TRIGGER update_value_ref
AFTER INSERT ON ref_table
FOR EACH ROW EXECUTE PROCEDURE update_value();

select create_reference_table('ref_table');
ERROR:  relation ref_table is not distributed

```

do we expect to see this error?

Reply:
> SaitTalhaNisanci 24 minutes ago Member
when I add && !IsReferenceTable(relationId) here, and do the following:

As the table is just being created, `IsReferenceTable` would error out as it is not a citus table yet. Proper way to test this can be to pass a flag `EnsureRelationHasNoTriggers`  to understand if we are creating a reference table.

Reply:
Right, I guess currently we get an error in this case as we don't handle triggers in `RelayEventExtendNames`. That is called from worker_apply_shard_ddl_command, and the error I get is the following:

```
talha=# select create_reference_table('ref_table');
DEBUG:  opening 1 new connections to localhost:9701
DEBUG:  opening 1 new connections to localhost:9702
DEBUG:  established connection to localhost:9702 for session 4
NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distrib
uted_transaction_id(0, 5, '2020-05-29 17:48:07.767295+03');
DETAIL:  on server talha@localhost:9702 connectionId: 2
NOTICE:  issuing SELECT worker_apply_shard_ddl_command (102205, 'CREATE TABLE public.re
f_table (value integer)');SELECT worker_apply_shard_ddl_command (102205, 'ALTER TABLE p
ublic.ref_table OWNER TO talha');SELECT worker_apply_shard_ddl_command (102205, 'CREATE
 TRIGGER ref AFTER INSERT ON public.ref_table FOR EACH ROW EXECUTE FUNCTION public.upda
te_value()')
DETAIL:  on server talha@localhost:9702 connectionId: 2
DEBUG:  unsafe statement type in name extension
DETAIL:  Statement type: 273
DETAIL:  from localhost:9702
ERROR:  relation "public.ref_table" does not exist
CONTEXT:  while executing command on localhost:9702
Time: 5.615 ms
```

Reply:
That's right, this is because `RelayEventExtendNames`does not know the way to extend `CREATE TRIGGER` commands for shard relations, which is required to issue necessary DDL commands (returned by `GetTableDDLEvents` ) on shards too.